core/schema/git.go

```diff
@@ -383,48 +388,80 @@ func (s *gitSchema) gitLegacy(ctx context.Context, parent dagql.Instance[*core.Q
 	})
 }
 
-func (s *gitSchema) head(ctx context.Context, parent *core.GitRepository, args struct{}) (*core.GitRef, error) {
-	return parent.Head(ctx)
-}
-
 type refArgs struct {
 	Name string
 }
 
-func (s *gitSchema) ref(ctx context.Context, parent *core.GitRepository, args refArgs) (*core.GitRef, error) {
-	return parent.Ref(ctx, args.Name)
+func (s *gitSchema) ref(ctx context.Context, parent dagql.Instance[*core.GitRepository], args refArgs) (inst dagql.Instance[*core.GitRef], _ error) {
+	result, err := parent.Self.Ref(ctx, args.Name)
+	if err != nil {
+		return inst, err
+	}
+	inst, err = dagql.NewInstanceForCurrentID(ctx, s.srv, parent, result)
+	if err != nil {
+		return inst, err
+	}
+
+	if ref, ok := inst.Self.Backend.(*core.RemoteGitRef); ok {
+		inst = inst.WithDigest(dagql.HashFrom(
+			// include the remote
+			ref.Repo.URL.Remote(),
+			// and the fully resolved ref + commit
+			ref.FullRef, ref.Commit,
+
+			// and the top-level discard git dir setting
+			strconv.FormatBool(inst.Self.Repo.DiscardGitDir),
+
+			// also include what auth methods are used, currently we can't
+			// handle a cache hit where the result has a different auth
+			// method than the caller used (i.e. a git repo is pulled w/
+			// a token but hits cache for a dir where a ssh sock was used)
+			strconv.FormatBool(ref.Repo.AuthToken.Self != nil),
+			strconv.FormatBool(ref.Repo.AuthHeader.Self != nil),
+			strconv.FormatBool(ref.Repo.SSHAuthSocket != nil),
```

Comment:
@sipsma so I think this is right.

Instead of having this be on `tree`, I've moved it up to the `ref` selection - so it'll apply to every operation on a `GitRef`.

The reasoning being, we should have a content hash *here*, which means we should easily get the following behavior:
- Cache the results of `dag().git().ref()` per session (which effectively means we only resolve each remote + ref combo once per session, similar to buildkit today).
- And then when it's not cached between sessions, get a "content hashed" call, that includes the ref + commit + remote + etc. So we don't need to invoke the actual op multiple times when this is shared.

But would really love a bit of a closer inspection, just to make sure I'm definitely thinking about this right.

Reply:
Uh yeah, I moved this back :thinking: I'm now slightly struggling to remember *exactly* why.

I think it was related to accidentally edge merging *way* earlier than we should be :thinking: but yes, ended up deciding to lift it up for now, just to entirely avoid that problem.